sharedbuffers/mapped_struct.py

```diff
@@ -1718,6 +1741,519 @@ def __str__(self):
+                    elif val == elem:
+                        return True
+                    mid -= 1
+
+                mid = lo + ((hi - lo) >> 1)
+                while mid < hi:
+                    val = self.objlist._c_getitem(mid, dcode, objlen, itemsize, offset, _struct, None)
+                    h1 = _stable_hash(val)
+                    if h1 != h2:
+                        break
+                    elif val == elem:
+                        return True
+                    mid += 1
+                return False
+            elif h1 < h2:
+                lo = mid + 1
+            else:
+                hi = mid
+        return False
+
+    @cython.locals(i=cython.Py_ssize_t, dcode=cython.char, objlen=cython.longlong,
+        itemsize=cython.uchar, offset=cython.Py_ssize_t)
+    def __iter__(self):
+        if self.objlist is not None:
+            dcode, objlen, itemsize, offset, _struct = self.objlist._metadata()
+            for i in xrange(len(self)):
+                yield self.objlist._c_getitem(i, dcode, objlen, itemsize, offset, _struct, None)
+        else:
+            if self.bitrep_lo:
+                for i in xrange(64):
+                    if self.bitrep_lo & (cython.cast(cython.ulonglong, 1) << i):
+                        yield i
+                    elif not self.bitrep_lo >> i:
+                        break
+
+            if self.bitrep_hi:
+                for i in xrange(64):
+                    if self.bitrep_hi & (cython.cast(cython.ulonglong, 1) << i):
+                        yield i + 64
+                    elif not self.bitrep_hi >> i:
+                        break
+
+    @cython.locals(i=cython.Py_ssize_t)
+    @cython.ccall
+    def _add_to(self, out):
+        if self.objlist is None:
+            for i in xrange(64):
+                if self.bitrep_lo & (cython.cast(cython.ulonglong, 1) << i):
+                    out.add(i)
+                elif not self.bitrep_lo >> i:
+                    break
+            for i in xrange(64):
+                if self.bitrep_hi & (cython.cast(cython.ulonglong, 1) << i):
+                    out.add(i + 64)
+                elif not self.bitrep_hi >> i:
+                    break
+            return out
+
+        dcode, objlen, itemsize, offset, _struct = self.objlist._metadata()
+        for i in xrange(len(self)):
+            out.add(self.objlist._c_getitem(i, dcode, objlen, itemsize, offset, _struct, None))
+        return out
+
+    @cython.ccall
+    @cython.locals(pfset='proxied_frozenset')
+    def _union_2(self, seq):
+        if not seq:
+            return self
+        elif self.objlist is None and type(seq) is proxied_frozenset:
+            pfset = cython.cast(proxied_frozenset, seq)
+            if pfset.objlist is None:
+                return proxied_frozenset(
+                    None, self.bitrep_lo | pfset.bitrep_lo, self.bitrep_hi | pfset.bitrep_hi)
+
+        rv = self._add_to(set())
+        rv.update(seq)
```

Comment:
I think it *should* be faster

``` python
return frozenset(self._add_to(set(seq)))
```

Because set copying is optimized in CPython (it just copies the hash table).

Reply:
This does seem to generate better code. Thanks.